File: modules/tests/src/commonTest/kotlin/primitives/ReturnPrimitive.kt
package primitives

import natives.test.MyDictionary
import natives.test.MyEnum
import withLib
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

class ReturnPrimitive {

    @Test
    fun returnVoid() = withLib {
        assertEquals(Unit, natives.test.returnVoid())
    }

    @Test
    fun returnChar() = withLib {
        assertEquals('a', natives.test.returnChar())
    }

    @Test
    fun returnBoolean() = withLib {
        assertEquals(true, natives.test.returnBoolean())
    }

    @Test
    fun returnByte() = withLib {
        assertEquals(99.toByte(), natives.test.returnByte())
    }

    @Test
    fun returnUByte() = withLib {
        assertEquals(UByte.MAX_VALUE, natives.test.returnUByte())
    }

    @Test
    fun returnShort() = withLib {
        assertEquals(99.toShort(), natives.test.returnShort())
    }

    @Test
    fun returnUShort() = withLib {
        assertEquals(UShort.MAX_VALUE, natives.test.returnUShort())
    }

    @Test
    fun returnInt() = withLib {
        assertEquals(99, natives.test.returnInt())
    }

    @Test
    fun returnUInt() = withLib {
        assertEquals(UInt.MAX_VALUE, natives.test.returnUInt())
    }

    @Test
    fun returnLong() = withLib {
        assertEquals(9223372036854775805L, natives.test.returnLong())
    }

    @Test
    fun returnULong() = withLib {
        assertEquals(ULong.MAX_VALUE, natives.test.returnULong())
    }

    @Test
    fun returnFloat() = withLib {
        assertEquals(99f, natives.test.returnFloat())
    }

    @Test
    fun returnDouble() = withLib {
        assertEquals(99.0, natives.test.returnDouble())
    }

    @Test
    fun returnString() = withLib {
        assertEquals("test string", natives.test.returnString())
    }

    @Test
    fun returnStringN() = withLib {
        assertNull(natives.test.returnStringN())
    }

    @Test
    fun returnEnum() = withLib {
        assertEquals(MyEnum.CASE2, natives.test.returnEnum())
    }

    @Test
    fun returnDictionary() = withLib {
        assertEquals(MyDictionary(1, 2, 3, 4), natives.test.returnDictionary())
    }

    @Test
    fun returnDictionaryN() = withLib {
        assertNull(natives.test.returnDictionaryN())
    }
}

File: modules/tests/natives/testcpp/src/primitives/return_primitive.cpp
#include <api.hpp>

void return_void() {
}

uint16_t return_char() {
    return 'a';
}

bool return_boolean() {
    return true;
}

int8_t return_byte() {
    return 99;
}

uint8_t return_ubyte() {
    return 255u;
}

int16_t return_short() {
    return 99;
}

uint16_t return_ushort() {
    return 65535u;
}

int32_t return_int() {
    return 99;
}

uint32_t return_uint() {
    return 4294967295u;
}

int64_t return_long() {
    return 9223372036854775805;
}

uint64_t return_ulong() {
    return 18446744073709551615u;
}

float return_float() {
    return 99;
}

double return_double() {
    return 99.0;
}

KString return_string() {
    return KString("test string");
}

KOptional<KString> return_string_n() {
    return KOptional<KString>();
}

MyEnum return_enum() {
    return CASE2;
}

MyDictionary return_dictionary() {
    return MyDictionary(1, 2, 3, 4);
}

KOptional<MyDictionary> return_dictionary_n() {
    return KOptional<MyDictionary>();
}